### DISCUSSION TITLE
Understanding mapfile-writer's zoom-interval-conf

### DISCUSSION THREAD

**Discussion by Just-Explode** [mapsforge/mapsforge#1463] [Q&A] [answered]
During my task of creating map for entire Taiwan using [kcwu's map](https://osm.kcwu.csie.org/download/tw-extract/recent/taiwan-latest.o5m) and osmconvert, mapfile-writer. I encountered a problem that indicates the default setting for zoom-interval-conf needed to be changed. I read the manual and see that i have to set base zoom and an interval, which I still don't know what is providing what. So in case of misunderstanding, I reconfigured it as 0,0,0,1,1,1,...,21,21,21,...

But I still wanna understand how these parameters work. For my current understanding. I think basezoom indicates the zoomlevel of input osm file and take data from it, while minZoom and maxZoom indicate the range of output map file zoomlevels. And the plugin will copy the selected basezoom from input file to the selected zoomlevel interval of output map file. Am I right?

And based on the assumptions, how do I know what zoomlevels are contained in the input osm file? And is the definitions of zoomlevels for input osm file and output map file the same as [OSM's website regarding zoom levels](https://wiki.openstreetmap.org/wiki/Zoom_levels)?

Btw here's my command rn
```
./osmosis --rx file=OSM-Map-Updater/Workstation/taiwan-latest.osm --mapfile-writer file=OSM-Map-Updater/OSM_Taiwan.map zoom-interval-conf=0,0,0,1,1,1,2,2,2,3,3,3,4,4,4,5,5,5,6,6,6,7,7,7,8,8,8,9,9,9,10,10,10,11,11,11,12,12,12,13,13,13,14,14,14,15,15,15,16,16,16,17,17,17,18,18,18 type=hd
```
Will append the numerous list if my pc can handle it
Now my phone gets a little be laggy when it's at around zoomlevel 13. So i might have to change the settings.

**Comment by devemux86** [comment, marked as answer]
Map writer documentation is [here](https://github.com/mapsforge/mapsforge/blob/master/docs/Getting-Started-Map-Writer.md).

**We should not change the default zoom interval.**

Zoom intervals have different simplification and contain different data based on zoom-appear in [tap-mapping](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map-writer/src/main/config/tag-mapping.xml).
To understand the structure of map file, you can study the [map file specification](https://github.com/mapsforge/mapsforge/blob/master/docs/Specification-Binary-Map-File.md).

**Reply by Just-Explode** [reply]
OMG thx for your help. i didn't realize that the entire mapsforge project is a map file format for mobile devices (im sorry). That gave me a whole new view of everything.

But now I feel more confused. So now I know the sub-files contain certain range of zoom level data(still not sure how the plugin gets the source). And each sub-file has its own supported zoomlevels range definded by minZoom and maxZoom. By the documentation, that explained why we we're recommended to have a consecutive and not overlapping interval.

And according to tag-mapping.xml, it seems that the plugin just pulls POI data from input no matter what and put them in the predefined zoom levels that's not higher than 17, in the asscoiated sub-file.
I assume the same thing also happens for ways or areas. So, no matter where data is stored in any zoom level in the original file, the plugin will replace them to appropriate zoom levels in the output file based on the definitions built within the plugin. Am I correct?

Assuming I was right, if I wanna make maximum use of the original file, all I have to do is to make sure I gave zoom-interval-conf the max range of zoomlevel that is supported by plugin's definitions. (Ex. If tag-mapping.xml says the highest zoom level it'll put any new data is 17, then the range of zoomlevels to be configured via zoom-interval-conf will be 0~17) There's no need to know how many zoom levels the input files have. Is that right?

But also, what's the purpose of defining baseZoom? If it's only for representing the levels in a sub-file, does that mean I can randomly choose a zoom level within as the basezoom number?
The only thing that stops me from doing it is this line of words in doc. There's nothing else related to basezoom:
```
Note: to calculate how many tile index entries there will be, use the formulae at [http://wiki.openstreetmap.org/wiki/Slippy_map_tilenames] to find out how many tiles will be covered by the bounding box at the base zoom level of the sub file
```
I haven't figured out the use of it, and I'm doing it now. But any help is appreciated!

Btw i do encountered a scenario that doesn't make the default and fallback config(according to doc regarding zoom-interval-conf) work.
Here's the error code. A result on the web that I can't recall says that zoom-interval-conf have to be modified.
```
嚴重: Thread for task 1-rx failed
java.nio.BufferOverflowException
        at java.base/java.nio.HeapByteBuffer.put(HeapByteBuffer.java:231)
        at java.base/java.nio.ByteBuffer.put(ByteBuffer.java:1210)
        at org.mapsforge.map.writer.MapFileWriter.processPOI(MapFileWriter.java:510)
        at org.mapsforge.map.writer.MapFileWriter.processTile(MapFileWriter.java:858)
        at org.mapsforge.map.writer.MapFileWriter.writeSubfile(MapFileWriter.java:974)
        at org.mapsforge.map.writer.MapFileWriter.writeFile(MapFileWriter.java:372)
        at org.mapsforge.map.writer.osmosis.MapFileWriterTask.complete(MapFileWriterTask.java:118)
        at org.openstreetmap.osmosis.xml.v0_6.XmlReader.run(XmlReader.java:53)
        at java.base/java.lang.Thread.run(Thread.java:840)

1月 27, 2024 6:59:37 下午 org.openstreetmap.osmosis.core.Osmosis main
嚴重: Execution aborted.
org.openstreetmap.osmosis.core.OsmosisRuntimeException: One or more tasks failed.
        at org.openstreetmap.osmosis.core.pipeline.common.Pipeline.waitForCompletion(Pipeline.java:146)
        at org.openstreetmap.osmosis.core.Osmosis.run(Osmosis.java:92)
        at org.openstreetmap.osmosis.core.Osmosis.main(Osmosis.java:37)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchStandard(Launcher.java:322)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:235)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:407)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:348)
        at org.codehaus.classworlds.Launcher.main(Launcher.java:47)
```
The Mandarin character doesn't matter. It just means ERROR.

Also what's the point of cutting sub-files? Is it for the program definitions of where the data is put?

**Reply by devemux86** [reply]
Mapsforge are vector maps, not bitmap tiles with fixed zoom levels.

_Do **not** set a custom zoom interval. Use the default, it's best for most purposes._

Also give Java / Osmosis more memory with `-Xmx` ([documentation](http://docs.oracle.com/javase/7/docs/technotes/tools/solaris/java.html)).
For large maps use the map-writer option `type=hd` so maps are not built in memory.

**Reply by Just-Explode** [reply]
Thanks for the information! I do see that the document says the file structure divided sub-files for different zoomlevels. But with your info, now I understand the reason why the zoom intervals were split in such way. So the zoom intervals contain data that is shown only after certain zoomlevels that users control by zoom in/out. The map file only remembers when it'll start appearing, definded by tag-mapping.xml. And it is split by the default way to prevent buffer memory blows up. That's why it's not recommended to modify it.
(Still doesn't see how the number of baseZoom definitions was defined here)

But without changing that intervals, the problem still exist.
I have tried `-Xmx51200M` and `type=hd` and both of them raised same error.
I have also tried to modify the zoom-interval-conf to another set as described by [this](https://github.com/mapsforge/mapsforge/blob/master/docs/Getting-Started-Map-Writer.md) manual.

The reason why I decided to change zoom-interval-conf is based on [this](https://github.com/mapsforge/mapsforge/issues/388#issuecomment-61954088)
discussion. Based on my current understanding, I should either modify tag-mapping.xml or zoom-interval-conf. However I'm working with a pre-built version of .jar file so idk if I can change the xml inside the jar directly.
So I'm now dealing with zoom intervals.
For now, 0,0,0,...,18,18,18 worked out. But 19 took too long(over 6 hours)

Another question is, why is the bottom of the intervals 21 instead of 17?

**Reply by devemux86** [reply]
_Use the default zoom interval_ and adjust tag-mapping to store detailed / complex data from zoom 12 (with zoom-appear).
The tag-mapping can be passed as external option in map-writer, see the map writer documentation.

You can also change the `Parameters.MAXIMUM_BUFFER_SIZE`, if you find current value small for your map reading.

**Reply by Just-Explode** [reply]
I wanna do a wild guess. Is basezoom related to vector simplification? If so, how does it work?
I see it's related to `simplification-factor`, `simplification-max-zoom`, `zoom-interval-conf`,[Conceptual design](https://github.com/mapsforge/mapsforge/blob/master/docs/Specification-Binary-Map-File.md#conceptual-design), [File header](https://github.com/mapsforge/mapsforge/blob/master/docs/Specification-Binary-Map-File.md#file-header) and [Tile index entry](https://github.com/mapsforge/mapsforge/blob/master/docs/Specification-Binary-Map-File.md#file-header)

**Reply by Just-Explode** [reply]
> _Use the default zoom interval_ and adjust tag-mapping to store detailed / complex data from zoom 12 (with zoom-appear).
> The tag-mapping can be passed as external option in map-writer, see the map writer documentation.
> 
> You can also change the `Parameters.MAXIMUM_BUFFER_SIZE`, if you find current value small for your map reading.

I'm ready to modity the tag-mapping.xml.

Where can I find MAXIMUM_BUFFER_SIZE in mapsforge/mapsforge-map-writer? I can't find it using search on github.

Should I be worried about increasing buffer size or I can just try it one step at a time assured because the discussion I found was from 10 years ago?

**Reply by devemux86** [reply]
It is used in map-reader ([here](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-core/src/main/java/org/mapsforge/core/util/Parameters.java#L38)).

**Reply by Just-Explode** [reply]
I have cloned the repository using `git clone https://github.com/mapsforge/mapsforge.git`
And ran the command `./gradlew clean build`
It raised this error
```

FAILURE: Build failed with an exception.

* Where:
Build file '/home/fred/OSM-Map-Updater/Workstation/tmp/mapsforge/mapsforge-map-android/build.gradle' line: 25

* What went wrong:
A problem occurred evaluating project ':mapsforge-map-android'.
> Environment variable ANDROID_HOME needs to be set to SDK folder

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 41s
```
However set it to my SDK location `/home/fred/Android/Sdk/` fixed it.
Should I report this?

Btw where should I find the built binary :v

**Reply by Just-Explode** [reply]
Hold on, forgive me, I missed the last line of the doc :/

**Reply by Just-Explode** [reply]
Welp, for some reason, the built version can generate the file without any error, without modifying zoom-interval-conf also. It's pretty weird, but I'll take it, thx for your help so far. I didn't see how obvious the method is.

The last problem is that, no matter the configuration of zoom-interval-conf is (default, (5,0,7,10,8,11,12,12,13,14,14,21),(0,0,0,...,18,18,18)), at around zoomlevel 8, the map start feel laggy.

I'm running it on OpenTrack's OSM Dashboard(Offline). I can still test it on Map Marker.

Edit: This problem doesn't exit on Map Marker. Maybe i should keep it as is?
The laggy scenario is like 21(smoothest)->12(a little bit laggy)->11(smooth again)->8(most laggy, but still controllable)->7(everything start feeling smooth from now on)->0
